File: Algorithms/Swift/Easy/206_Reverse_Linked_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func reverseList(head: ListNode?) -> ListNode? {
        if head == nil {
            return nil
        }

        var prev: ListNode? = nil
        var node = head
        while node != nil {
            let tmp = node!.next;
            node!.next = prev
            prev = node
            node = tmp
        }

        return prev
    }
}

File: Algorithms/C++/Easy/206_Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Iteration
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;

        while (head != NULL) {
            ListNode* tmp = head->next;
            head->next = prev;
            prev = head;
            head = tmp;
        }

        return prev;
    }
};

// Recursion
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }

        ListNode* newTail = head->next;

        ListNode* newHead = reverseList(head->next);

        newTail->next = head;
        head->next = NULL;

        return newHead;
    }
};
